util/src/main/java/tech/pegasys/artemis/util/async/SafeFuture.java

```diff
@@ -96,6 +97,27 @@ public static void reportExceptions(final CompletionStage<?> future) {
     return of(CompletableFuture.allOf(futures));
   }
 
+  /**
+   * Returns a new SafeFuture that is completed when all of the given SafeFutures complete
+   * successfully or completes exceptionally immediately when any of the SafeFutures complete
+   * exceptionally. The results, if any, of the given SafeFutures are not reflected in the returned
+   * SafeFuture, but may be obtained by inspecting them individually. If no SafeFutures are
+   * provided, returns a SafeFuture completed with the value {@code null}.
+   *
+   * <p>Among the applications of this method is to await completion of a set of independent
+   * SafeFutures before continuing a program, as in: {@code SafeFuture.allOf(c1, c2, c3).join();}.
+   *
+   * @param futures the SafeFutures
+   * @return a new SafeFuture that is completed when all of the given SafeFutures complete
+   * @throws NullPointerException if the array or any of its elements are {@code null}
+   */
+  public static <U> SafeFuture<Void> allOfFailFast(final SafeFuture<?>... futures) {
+    final SafeFuture<Void> complete = new SafeFuture<>();
+    Stream.of(futures).forEach(future -> future.finish(() -> {}, complete::completeExceptionally));
```

Comment:
Would it be worth using a parallel stream rather than a sequential stream?  
```
Stream.of(futures). parallel().forEach(future -> future.finish(() -> {}, complete::completeExceptionally));
```

Reply:
No definitely not.  It adds a lot of complexity to reason about and the overhead of cross-thread communication is not going to be close to worth it for what is always a very small number of futures (and calling `finish` on each is a very cheap operation).

It's very rare that `Stream.parallel` is a good idea. :)

Reply:
Fair enough